services/discovery/discovery.go

```diff
@@ -0,0 +1,50 @@
+// Package discovery implements the discovery service, which lets users surface resource configs for their machines to use.
+package discovery
+
+import (
+	"context"
+
+	pb "go.viam.com/api/service/discovery/v1"
+
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+)
+
+func init() {
+	resource.RegisterAPI(API, resource.APIRegistration[Service]{
+		RPCServiceServerConstructor: NewRPCServiceServer,
+		RPCServiceHandler:           pb.RegisterDiscoveryServiceHandlerFromEndpoint,
+		RPCServiceDesc:              &pb.DiscoveryService_ServiceDesc,
+		RPCClient:                   NewClientFromConn,
+	})
+}
+
+// SubtypeName is the name of the type of service.
+const (
+	SubtypeName = "discovery"
+)
+
+// API is a variable that identifies the slam resource API.
+var API = resource.APINamespaceRDK.WithServiceType(SubtypeName)
+
+// Named is a helper for getting the named service's typed resource name.
+func Named(name string) resource.Name {
+	return resource.NewName(API, name)
+}
+
+// FromRobot is a helper for getting the named discovery service from the given Robot.
+func FromRobot(r robot.Robot, name string) (Service, error) {
+	return robot.ResourceFromRobot[Service](r, Named(name))
+}
+
+// FromDependencies is a helper for getting the named discovery service from a collection of
+// dependencies.
+func FromDependencies(deps resource.Dependencies, name string) (Service, error) {
+	return resource.FromDependencies[Service](deps, Named(name))
+}
+
+// Service describes the functions that are available to the service.
+type Service interface {
+	resource.Resource
+	DiscoverResources(ctx context.Context, extra map[string]any) ([]*resource.Config, error)
```

Comment:
Does this need to return an array of *pointers* to resource.Configs? Or would it be okay to return `[]resource.Config`, since you can already return nil here?

Reply:
I'm not familiar with this project but in terms of API design:

At a cursory glance, the vast majority of `resource.Config`s are passed and returned from functions as structs rather than pointers to structs.

In my opinion, as long as `DiscoverResources` isn't performance sensitive (aka we can afford to copy all the `resource.Config`s), we should return   `[]resource.Config` rather than `[]*resource.Config` to be consistent with the other uses of `resource.Config` and b/c it makes it clear that the caller is getting a copy of the data which they are allowed to modify if they choose.

Reply:
The API is meant to aid users in configuring their machines - it can return a large array - say you're configuring a thousand cameras at once - but I don't see it needing to be that performant - and I think arrays in go are pointers under the hood anyway?

Reply:
Slices are indeed pointers in go. However each element of the `[]resource.Config` is going to have a new copy of `resource.Config`, where as `[]*resource.Config` can return pointers to the same objects rather than needing to make copies. As a result, if the slice has a lot of elements and the `resource.Config` is very big I would expect `[]*resource.Config` to consume less memory than `[]resource.Config`. 

That said, returning pointers to objects that other parts of the system are using (and don't expect to change) is a common source of bugs, so I think we should prefer the safer and slightly more expensive method signature until we know the performance matters. I agree that this code is unlikely to be very performance-sensitive.